tensorstore/kvstore/s3/s3_key_value_store_test.cc

```diff
@@ -0,0 +1,125 @@
+// Copyright 2020 The TensorStore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stddef.h>
+
+#include <algorithm>
+#include <string>
+#include <string_view>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include "absl/log/absl_log.h"
+#include "absl/status/status.h"
+#include "absl/strings/match.h"
+#include "absl/synchronization/mutex.h"
+#include "absl/synchronization/notification.h"
+#include "absl/time/clock.h"
+#include "absl/time/time.h"
+#include <nlohmann/json.hpp>
+#include "tensorstore/context.h"
+#include "tensorstore/internal/http/curl_handle.h"
+#include "tensorstore/internal/http/curl_transport.h"
+#include "tensorstore/internal/http/http_request.h"
+#include "tensorstore/internal/http/http_response.h"
+#include "tensorstore/internal/json_gtest.h"
+#include "tensorstore/internal/path.h"
+#include "tensorstore/internal/schedule_at.h"
+#include "tensorstore/kvstore/generation.h"
+#include "tensorstore/kvstore/kvstore.h"
+#include "tensorstore/kvstore/operations.h"
+#include "tensorstore/kvstore/test_util.h"
+#include "tensorstore/util/execution/execution.h"
+#include "tensorstore/util/execution/sender_testutil.h"
+#include "tensorstore/util/executor.h"
+#include "tensorstore/util/future.h"
+#include "tensorstore/util/status_testutil.h"
+#include "tensorstore/util/str_cat.h"
+
+namespace {
+
+namespace kvstore = ::tensorstore::kvstore;
+
+using ::tensorstore::CompletionNotifyingReceiver;
+using ::tensorstore::Context;
+using ::tensorstore::Future;
+using ::tensorstore::KeyRange;
+using ::tensorstore::MatchesJson;
+using ::tensorstore::MatchesStatus;
+using ::tensorstore::StorageGeneration;
+using ::tensorstore::StrCat;
+using ::tensorstore::internal::ScheduleAt;
+using ::tensorstore::internal_http::HttpRequest;
+using ::tensorstore::internal_http::HttpResponse;
+using ::tensorstore::internal_http::HttpTransport;
+using ::tensorstore::internal_http::SetDefaultHttpTransport;
+
+static constexpr char kUriScheme[] = "s3";
+static constexpr char kDriver[] = "s3";
+
+Context DefaultTestContext() {
```

Comment:
So, we probably need a mock/fake s3 server for testing. You could implement a simple one yourself, like the GCS mock we use in tensorstore/kvstore/gcs/gcs_mock.h

Or you might be able to use a python variant or some other one, like:  https://github.com/findify/s3mock
Coupled with a process spawning mechanism (for exposition only):


```
//  S3Mock runs the mock process and returns a port.
class S3Mock {
 public:
   S3Mock();
  void SpawnProcess();
  std::string http_endpoint() { return tensorstore::StrCat("localhost:", port); }
  
  const int port;
  std::optional<tensorstore::internal::Subprocess> child;
};

S3Mock::S3Mock() : port(TryPickUnusedPort().value_or(0)) {}

void S3Mock::SpawnProcess() {
  if (child.has_value()) return;

  // Spawn the process.
  TENSORSTORE_CHECK_OK_AND_ASSIGN(
    child, SpawnSubprocess(SubprocessOptions{"path/to/mock", {tensorstore::StrCat("--port=", port)}}));

  // Wait for connection
  for (auto deadline = absl::Now() + absl::Seconds(10);;) {
    absl::SleepFor(absl::Milliseconds(100));
    auto ready_future = GetDefaultHttpTransport()->IssueRequest(
      HttpRequestBuilder("GET",  tensorstore::StrCat("http://localhost:", port, "/startup")).BuildRequest(),
      absl::Cord(), absl::Seconds(10), absl::Seconds(10));
    if (ready_future.status().ok()) break;
    if (absl::Now() < deadline) continue;
    TENSORSTORE_CHECK_OK(ready_future.status());
  }
}

```

Reply:
Another option could be minio.  Google unfortunately has a policy against AGPL code, but we could use minio 2021-04-22 (https://github.com/minio/minio/tree/RELEASE.2021-04-22T15-44-28Z) which is the last release licensed under Apache 2.

Reply:
Aiming to work on an S3 Mock next, now that things are working with the AWS S3 Servers.

Any thoughts on [localstack](https://github.com/localstack/localstack)? Looks like it needs to run in a container though.

Reply:
Looks like https://github.com/findify/s3mock doesn't support signature checking.

I've used minio for testing before so minio 2021-04-22 might be the easiest.

Reply:
FWIW: Now we use gcs testbench in several of our tests using something similar to the above.

https://github.com/google/tensorstore/blob/master/tensorstore/kvstore/gcs/gcs_testbench.h

Reply:
Added a localstack fixture